### PR TITLE
Blahgger

### PR DESCRIPTION
**Pull request by jbvera**
Basic authentication done

### CODE REVIEW

File: Blahgger/Blahgger/Data/BlahggerData.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        private BlahggerData() { }
+
+        public static BlahggerData GetInstance()
+        {
+            if (singleton == null)
+            {
+                singleton = new BlahggerData();
+            }
+            return singleton;
+        }
```
Comment (jbvera) [inline Blahgger/Blahgger/Data/BlahggerData.cs]: Sorry about the singleton data class. This was before our conversation about stateless programming.

File: Blahgger/Blahgger/Controllers/AccountsController.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+using System.Linq;
+using System.Web;
+using System.Web.Mvc;
+
+namespace Blahgger.Controllers
+{
+    public class AccountsController : Controller
+    {
+        // GET: Accounts/Create
+        public ActionResult Create()
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Controllers/AccountsController.cs]: Should this action method allow anonymous users?

File: Blahgger/Blahgger/Controllers/AccountsController.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            User user = new User();
+            return View(user);
+        }
+
+        // POST: Accounts/Create
+        [HttpPost]
+        public ActionResult Create(User user)
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Controllers/AccountsController.cs]: Should this action method allow anonymous users?

File: Blahgger/Blahgger/Controllers/PostsController.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            BlahggerData data = BlahggerData.GetInstance();
+            List<Post> posts = data.GetPosts();
+            PostsViewModel postsViewModel = new PostsViewModel() { Posts = posts };
+            
+            return View(postsViewModel);
+        }
+
+        // GET: Posts/Details/5
+        public ActionResult Details(int id)
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Controllers/PostsController.cs]: Is this where you left off? Maybe add a TODO comment here?

File: Blahgger/Blahgger/Data/BlahggerData.cs (hunk 1 above)
Comment (appacademy-jchurchill) [reply Blahgger/Blahgger/Data/BlahggerData.cs]: No worries at all! Thanks for the reminder comment :)

File: Blahgger/Blahgger/Data/BlahggerData.cs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            ";
+            // TODO Password hashing
+
+            DatabaseHelper.Insert(sql,
+                new SqlParameter("@Username", user.Username),
+                new SqlParameter("@HashedPassword", user.HashedPassword));
+
+        }
+
+        internal int GetUsersId(string username)
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Data/BlahggerData.cs]: Is the use of `public` and `internal` access modifiers intentional in this class?

File: Blahgger/Blahgger/Data/BlahggerData.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+            DataTable userTable = DatabaseHelper.Retrieve(sql,
+                new SqlParameter("@Username", username));
+            if (userTable.Rows.Count == 1)
+            {
+                return userTable.Rows[0].Field<int>("Id");
+            }
+            else
+            {
+                throw new Exception("SELECT query returned unexpected number of rows");
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Data/BlahggerData.cs]: Nice data quality check!

File: Blahgger/Blahgger/Data/BlahggerData.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            }
+        }
+
+        /// <summary>
+        /// This method checks whether or not a user's credentials match a row in the database.
+        /// </summary>
+        /// <param name="user">The input user object. This is assumed to have non-null Username and HashedPassword properties.</param>
+        /// <returns>True if a match is found; false otherwise.</returns>
+        internal bool AuthenticateUser(User user)
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Data/BlahggerData.cs]: I'd separate the retrieval of the user's hashed password from the authentication-related code. Either do the authentication portion in the controller or add another security-related class.

File: Blahgger/Blahgger/Global.asax.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    public class MvcApplication : System.Web.HttpApplication
+    {
+        protected void Application_Start()
+        {
+            AreaRegistration.RegisterAllAreas();
+            RouteConfig.RegisterRoutes(RouteTable.Routes);
+
+            // Add global filters
+            GlobalFilters.Filters.Add(new HandleErrorAttribute());
+            GlobalFilters.Filters.Add(new System.Web.Mvc.AuthorizeAttribute());
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Global.asax.cs]: Nice job adding the `Authorize` attribute globally.

File: Blahgger/Blahgger/Views/Accounts/Create.cshtml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+                @Html.TextBoxFor(model => model.Username, new { htmlAttributes = new { @class = "form-control" } })
+                @Html.ValidationMessageFor(model => model.Username, "", new { @class = "text-danger" })
+            </div>
+        </div>
+
+        <div class="form-group">
+            @Html.LabelFor(model => model.HashedPassword, htmlAttributes: new { @class = "control-label col-md-2" })
+            <div class="col-md-10">
+                @Html.PasswordFor(model => model.HashedPassword, new { htmlAttributes = new { @class = "form-control" } })
+                @Html.ValidationMessageFor(model => model.HashedPassword, "", new { @class = "text-danger" })
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Views/Accounts/Create.cshtml]: I think we talked about this in class after you wrote this, but make sure that the user's hashed password doesn't make it to the view by using two models: one for interacting with the database and another for passing data to the view.

File: Blahgger/Blahgger/Views/Accounts/Create.cshtml [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+@{
+    ViewBag.Title = "Create";
+}
+
+<h2>Create</h2>
+
+@using (Html.BeginForm()) 
+{
+    @Html.AntiForgeryToken()
```
Comment (appacademy-jchurchill) [inline Blahgger/Blahgger/Views/Accounts/Create.cshtml]: Don't forget to add the `ValidateAntiForgeryToken` attribute to your POST action method.

**Pull request merged**